### PR TITLE
Organs now freeze with atmosphere and core temperature

### PR DESCRIPTION
**Pull request by cacogen**
## About The Pull Request

- Organs are frozen by 0 celsius atmosphere
- Organs are frozen in human mobs with core temperatures beneath 0 celsius AND beneath their species' cold damage limit
- Frozen organs in living mobs are now failing organs and will kill their owner if the owner doesn't warm up
- Frozen organs now look visually frozen and prepend the word 'frozen' to their names, like other frozen objects
- Human mobs with skin (not core) temperatures beneath their species' cold damage limit now look paler
- Human mobs frozen by morgue trays, freezer crates and freezer closets now look paler, to indicate their organs are frozen
- Can't do surgeries with organ_to_manipulate set if that organ is frozen
- Can't do surgeries if the patient is frozen (i.e. core temperature beneath both 0 celsius and their species' cold damage limit)

## Why It's Good For The Game

- Organs and bodies freezing in cold atmosphere creates more opportunities to preserve them, useful with an overrun medbay
- Making mobs pale when they're taking cold damage indicates it to other players
- Making mobs pale when their organs have been frozen by a container makes organ preservation in bodies more obvious 
- Making organs visually frozen makes organ preservation inside and outside of bodies obvious instead of having no easy way of knowing it's occurring
- Preventing surgeries on frozen mobs and frozen organs prevents using atmosphere to get around organ decay on dead mobs on operating tables which would be too easy

Todo
- [ ] Make morgue trays, freezer closets and freezer crates lower mobs' core temperatures to eventually freeze their organs, instead of doing it after a short delay and without regard for core temperature
- [ ] Unfreeze organs when entering cryo to prevent further organ damage to them, given cryo doesn't freeze them and who really wants to make cryo shittier
- [ ] Test temperature guns and cryokinesis
- [ ] Consider balancing advanced surgery on mobs with frozen organs (i.e. non-decaying organs on an operating table in a cold environment) using the slowdown cryostylane uses instead of preventing it entirely

## Changelog

:cl:
balance: Organs now freeze in 0 celsius atmosphere
balance: Organs in human mobs freeze when the mob's core temperature is beneath 0 celsius AND their species' cold damage limit
balance: Frozen organs in living mobs are failing organs and will kill their owner if they don't warm up
add: Frozen organs are now visibly frozen
add: Human mobs now look paler when their skin temperature is beneath their species' cold damage limit
add: Human mobs now look paler when their organs are frozen by structures like morgue trays, freezer closets and freezer crates
add: Can't do surgeries on frozen bodies or when the surgery is targeting a specific organ and that organ is frozen

### CODE REVIEW

**Comment by GuillaumePrata** [comment]
Plasmamen's base body temperature is -3C, how would it work for them?

**Comment by SuperNovaa41** [comment]
> Can't do surgeries with a set organ to manipulate if that organ is frozen or do surgeries on a dead body if it's frozen

This isn't mentioned in your changelog and its arguably the most important thing to be mentioned

**Comment by cacogen** [comment]
> Plasmamen's base body temperature is -3C, how would it work for them?

Thanks for bringing this to my attention. 

It doesn't cause their organs to freeze or prevent surgery, because these only happen when the mob is dead. 

When dead, their bodies are no longer pulling them down to -3 celsius so you'll be able to do surgeries on them.  

However, they do say they have pale skin when examined, which is something I'll have to fix. 

This also made me realise there's nothing to make organs freeze on death if the body is already freezing, so that's something I'll have to fix too.

**Comment by cacogen** [comment]
I've changed it and now the pale skin message doesn't reference skin (plasmamen don't have skin), and it only shows up when the human mob is beneath their species' cold damage limit (i.e. is taking cold damage). This means plasmamen won't have the message by default, due to having core temperatures beneath 0 celsius.

I've also made it so a body can freeze while alive (making it pale and freezing its organs and preventing surgery on it), but only if its core temperature is beneath 0 celsius and its species' cold damage limit. This means you will be able to do surgery on plasmamen as usual and they won't be walking around with frozen organs by default.

**Comment by cacogen** [comment]
I added preventing surgery on frozen bodies because when I saw Cobby assigned, based on forum threads about advanced surgeries and stasis beds they've posted in, I figured they'd be against having an easy way to do advanced surgeries without having to worry about organ decay. 

It's fairly easy to make a small room cold enough to freeze corpses with this (particularly on Ice Box, it's not like corpses breathe), while you wear a winter coat to protect you against damage.

>does having an immunity to cold (such as from, say, the temperature immunity mutation) make you unaffected by this? what about being in a cryo tube? a stasis bed?

Temperature adaptation gives you the RESIST_COLD trait which prevents the block of code in body_temperature_damage() from running where the ON_ICE trait is added (which freezes your organs) if your core temperature is low enough.

Cryo tubes stop body temperature from regulating altogether, so organs won't freeze if you're in one. They won't unfreeze either, which I'll have to figure out what to do about.

Stasis beds stop body_temperature_damage() from running when you're in one, so organs won't freeze. They won't unfreeze either, which I'll have to figure out what to do about.

>will temp guns, cryokinesis, and/or changeling cryo stings become incredibly lethal because of this? what's this threshold's relation to the cryo pistol's victim detonation threshold?

I tested the changeling cryo sting, and if you stacked it about 4 times on a human their organs failed after about a minute and a half and by the two minute mark they were dead. 

That made me revert frozen organs being failing organs and just make it so that frozen organs take organ damage of the same amount as EMP'd cybernetic organs (which is an amount that could be changed based on balance needs).

I'll have to test temperature guns and cryokinesis to see what their effects are.

Thank you for this feedback.

>if you really, really think that using a cold room instead of formaldehyde to preserve corpses you're performing surgery on is a gamebreaking issue, you can just multiply surgery lengths by 1.5x (or whatever penalty cryostylane uses) if the subject is cold instead of blocking them altogether.

I could look into doing that instead.

**Comment by ShadowAKT** [comment]
How would this work with cybernetic organs? It would make sense for the tiers above basic to be immune to cold, at the very least.

What if you were to transplant a plasmaman's organ into a human? Would the organ lose its cold immunity when it wouldn't make sense to?

Also would cryo cause organ death, since it lowers body temp? I would imagine you'd have to change cryoxadone to prevent this.

**Comment by cacogen** [comment]
>How would this work with cybernetic organs? It would make sense for the tiers above basic to be immune to cold, at the very least.

They freeze like regular organs in certain cases, but not others. They don't freeze on cold turfs, but they do freeze if they're inside a body in a morgue tray, for example. I'll make it consistent.

>What if you were to transplant a plasmaman's organ into a human? Would the organ lose its cold immunity when it wouldn't make sense to?

Their organs don't have cold immunity. Organs freeze in human mobs when their core temperature is both below 0 celsius and below that mob's threshold for taking cold damage. Plasmamen and humans have different thresholds for taking cold damage. Plasmamen's is much lower than humans'.

>Also would cryo cause organ death, since it lowers body temp?

The organs can't freeze because a cryo check stops natural body temperature regulation, which prevents the bit of code that causes cold damage from running (which is where the code that causes organs to freeze if the mob's core temperature is below the cold damage threshold is).

File: code/__HELPERS/icons.dm [hunk 1]
```diff
@@ -1055,8 +1055,57 @@ GLOBAL_LIST_EMPTY(friendly_animal_types)
...
+	SIGNAL_HANDLER
+	addtimer(CALLBACK(src, .proc/make_frozen), 3 SECONDS)
+
+/atom/proc/begin_unfreeze()
+	SIGNAL_HANDLER
+	if(HAS_TRAIT(src, TRAIT_ON_ICE)) //something is freezing us, don't unfreeze
+		return
+	addtimer(CALLBACK(src, .proc/make_unfrozen), 3 SECONDS)
+
+/atom/proc/make_frozen()
```
Comment (Kylerace) [inline code/__HELPERS/icons.dm:1068]: doc

File: code/__HELPERS/icons.dm [hunk 2]
```diff
@@ -1055,8 +1055,57 @@ GLOBAL_LIST_EMPTY(friendly_animal_types)
 
 GLOBAL_LIST_INIT(freon_color_matrix, list("#2E5E69", "#60A2A8", "#A1AFB1", rgb(0,0,0)))
 
-/obj/proc/make_frozen_visual()
-	// Used to make the frozen item visuals for Freon.
+/atom/proc/begin_freeze()
+	SIGNAL_HANDLER
+	addtimer(CALLBACK(src, .proc/make_frozen), 3 SECONDS)
+
+/atom/proc/begin_unfreeze()
```
Comment (Kylerace) [inline code/__HELPERS/icons.dm:1062]: doc

File: code/__HELPERS/icons.dm [hunk 3]
```diff
@@ -1055,8 +1055,57 @@ GLOBAL_LIST_EMPTY(friendly_animal_types)
 
 GLOBAL_LIST_INIT(freon_color_matrix, list("#2E5E69", "#60A2A8", "#A1AFB1", rgb(0,0,0)))
 
-/obj/proc/make_frozen_visual()
-	// Used to make the frozen item visuals for Freon.
+/atom/proc/begin_freeze()
```
Comment (Kylerace) [inline code/__HELPERS/icons.dm:1058]: doc

File: code/__HELPERS/icons.dm [hunk 4]
```diff
@@ -1055,8 +1055,57 @@ GLOBAL_LIST_EMPTY(friendly_animal_types)
...
+/atom/proc/begin_unfreeze()
+	SIGNAL_HANDLER
+	if(HAS_TRAIT(src, TRAIT_ON_ICE)) //something is freezing us, don't unfreeze
+		return
+	addtimer(CALLBACK(src, .proc/make_unfrozen), 3 SECONDS)
+
+/atom/proc/make_frozen()
+	return
+
+/atom/proc/make_unfrozen()
```
Comment (Kylerace) [inline code/__HELPERS/icons.dm:1071]: doc

File: code/__HELPERS/icons.dm (hunk 3 above)
Comment (Kylerace) [reply code/__HELPERS/icons.dm:1058]: also all of these should go into their type paths relevant file like the atom ones to the atom level and the carbon ones to the carbon level

File: code/__HELPERS/icons.dm [hunk 5]
```diff
@@ -1055,8 +1055,57 @@ GLOBAL_LIST_EMPTY(friendly_animal_types)
...
+	if(HAS_TRAIT(src, TRAIT_ON_ICE)) //something is freezing us, don't unfreeze
+		return
+	for(var/obj/item/organ/O in internal_organs)
+		REMOVE_TRAIT(O, TRAIT_ON_ICE, "owner body temp")
+		O.make_unfrozen() //no delay
+
+/mob/living/carbon/human/make_frozen()
+	if(!HAS_TRAIT(src, TRAIT_ON_ICE))
+		return
+	if(HAS_TRAIT_FROM(src, TRAIT_COLD_SKIN, "make_frozen()")) //already have it, don't need to apply it again
```
Comment (Kylerace) [inline code/__HELPERS/icons.dm:1091]: dont name trait sources after functions like this its hard to work with. even if that source is unique to that proc it should be something like "frozen solid" or whatever

File: code/__HELPERS/icons.dm (hunk 3 above)
Comment (Kylerace) [reply code/__HELPERS/icons.dm:1058]: also why are these on /atom level if theyre only used by /atom/movable

File: code/modules/mob/living/carbon/human/species.dm [hunk 6]
```diff
@@ -1672,6 +1673,19 @@ GLOBAL_LIST_EMPTY(features_by_species)
...
 
+	var/is_hulk = HAS_TRAIT(humi, TRAIT_HULK)
+
+	var/cold_damage_limit = bodytemp_cold_damage_limit + (is_hulk ? BODYTEMP_HULK_COLD_DAMAGE_LIMIT_MODIFIER : 0)
+
+	//give body blue/grey hue if we're beneath the cold damage limit
+	if(humi.bodytemperature < cold_damage_limit && !HAS_TRAIT(humi, TRAIT_RESISTCOLD))
+		if(HAS_TRAIT_FROM(humi, TRAIT_COLD_SKIN, "body temp"))
+			return
+		ADD_TRAIT(humi, TRAIT_COLD_SKIN, "body temp")
```
Comment (Kylerace) [inline code/modules/mob/living/carbon/human/species.dm:1684]: trait sources like these should be defines

File: code/modules/ruins/lavalandruin_code/biodome_winter.dm [hunk 7]
```diff
@@ -42,7 +42,9 @@
 			hit_object.visible_message(span_warning("[hit_object] is freeze-proof!"))
 			return
 		if(!(hit_object.obj_flags & FROZEN))
-			hit_object.make_frozen_visual()
+			ADD_TRAIT(hit_object, TRAIT_ON_ICE, type)
+			hit_object.make_frozen() //immediately freeze
```
Comment (Kylerace) [inline code/modules/ruins/lavalandruin_code/biodome_winter.dm:46]: why do you need to add a trait to an object just to call a proc and then remove the trait

File: code/modules/surgery/organs/organ_internal.dm [hunk 8]
```diff
@@ -51,6 +51,9 @@ INITIALIZE_IMMEDIATE(/obj/item/organ)
 			foodtypes = RAW | MEAT | GROSS,\
 			volume = reagent_vol,\
 			after_eat = CALLBACK(src, .proc/OnEatFrom))
+
+	RegisterSignal(src, SIGNAL_ADDTRAIT(TRAIT_ON_ICE), /atom.proc/begin_freeze)
```
Comment (Kylerace) [inline code/modules/surgery/organs/organ_internal.dm:55]: this is a bad design pattern its confusing as to what the future programmer who has to work with your code is doing. youre essentially using adding this trait to an organ is a proc call which isnt clear at all unless someone finds these two lines.

**Review by Kylerace — CHANGES_REQUESTED** [review]
im wary of this. not only are you adding to two parts of the Life() loop youre attaching this behavior to every organ in a carbon and im not sure what the benefits over the frozen organ flag this has

**Comment by cacogen** [comment]
Is the Life() loop not the right place to do it?

I made it a trait because there can be multiple things freezing an organ simultaneously and I needed to keep track of those things to prevent organs unfreezing prematurely when one is removed.

By making it use the obj_flag it means that other sources of freezing also prevent an organ's decay, which is important for consistency when making frozen organs look like other frozen objects.

File: code/modules/ruins/lavalandruin_code/biodome_winter.dm (hunk 7 above)
Comment (cacogen) [reply code/modules/ruins/lavalandruin_code/biodome_winter.dm:46]: I don't like it either but the proc doesn't work without the trait. How would you do it?

File: code/modules/surgery/organs/organ_internal.dm (hunk 8 above)
Comment (cacogen) [reply code/modules/surgery/organs/organ_internal.dm:55]: So, remove the signal and call the proc directly?

**Comment by cacogen** [comment]
It was not a good idea to have freezing organs fail, because it made things that lowered body temperature drastically like the cryo sting ridiculously OP. Instead now they just take damage like EMP'd cybernetic ones.

It's quite difficult to get your core temperature that low (at least as a human). When I tested this I needed to be in like a -50 celsius room without any clothes for it to happen.

**Comment by LemonInTheDark** [comment]
I keep meaning to fix that

**Pull request closed without merging**